host_source.go

```diff
@@ -721,3 +732,160 @@ func (r *ringDescriber) refreshRing() error {
 	r.session.policy.SetPartitioner(partitioner)
 	return nil
 }
+
+const (
+	ringRefreshDebounceTime = 1 * time.Second
+)
+
+// debounces requests to call a refresh function (currently used for ring refresh). It also supports triggering a refresh immediately.
+type refreshDebouncer struct {
+	mu           sync.Mutex
+	stopped      bool
+	broadcaster  *errorBroadcaster
+	interval     time.Duration
+	timer        *time.Timer
+	refreshNowCh chan struct{}
+	quit         chan struct{}
+	refreshFn    func() error
+}
+
+func newRefreshDebouncer(interval time.Duration, refreshFn func() error) *refreshDebouncer {
+	d := &refreshDebouncer{
+		stopped:      false,
+		broadcaster:  nil,
+		refreshNowCh: make(chan struct{}),
```

Comment:
It seems that refreshNowCh should have capacity one instead of zero. The sender does non-blocking send, so with capacity zero, there is a chance that the refresh signal is lost when there is no receiver (for example, while refreshFn is running or while errorBroadcaster.broadcast is running).

Reply:
Oh nice catch, I missed this. I've changed this to have a capacity of 1.